### PR TITLE
feat(query): tail-array optimization in aggregate partial

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

If the reduce_factor is too low, we use tail-array in hashtable via `enable_tail_array` method.

Closes #issue

### CODE REVIEW

File: src/common/hashtable/src/tail_array.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+use std::alloc::Allocator;
+
+use super::table0::Entry;
+use super::traits::Keyable;
+use crate::container::Container;
+use crate::container::StackContainer;
+
+const SIZE: usize = 8192;
+
+pub struct TailArray<K, V, A: Allocator> {
```
Comment (BohuTANG) [inline src/common/hashtable/src/tail_array.rs:24]: It's better to add unit test for the tail array.

File: src/common/hashtable/src/tail_array.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    pub fn new(allocator: A) -> Self {
+        Self {
+            datas: vec![],
+            num_items: 0,
+            allocator,
+        }
+    }
+
+    pub fn insert(&mut self, key: K) -> &mut Entry<K, V> {
+        let pos = self.num_items % SIZE;
```
Comment (zhang2014) [inline src/common/hashtable/src/tail_array.rs:44]: tail array is ring buffer?

Comment (sundy-li) [reply src/common/hashtable/src/tail_array.rs:44]: No, it's a `Vector<  [Entrys; N]  >`. 

If we enable the tail array, the new keys will be pushed into this vector without expanding the hashtable.

**Comment by sundy-li** [comment]
Improved a little.


image


Memory went down fast after running the whole bench scripts.

image

File: src/common/hashtable/src/tail_array.rs (hunk 2 above)
Comment (zhang2014) [reply src/common/hashtable/src/tail_array.rs:42]: Is it only for the last block?

File: src/query/service/src/pipelines/processors/transforms/aggregator/aggregator_twolevel.rs [hunk 3]
```diff
@@ -213,8 +227,11 @@ where
             return Ok(data_blocks);
         }
 
-        drop(agg.area.take());
-        agg.states_dropped = true;
+        if agg.pass_state_to_final {
```
Comment (zhang2014) [inline src/query/service/src/pipelines/processors/transforms/aggregator/aggregator_twolevel.rs:230]: drop area if agg.pass_state_to_final == false?

**Pull request closed without merging**